### PR TITLE
Using GitOps with Grafana Operator to create Grafana Dashboard with Flux.

### PR DESCRIPTION
**Pull request by elamaran11**
### What does this PR do?

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted. Consult the CONTRIBUTING guide for submitting pull-requests.

### Motivation

Using GitOps with Grafana Operator to create Grafana Dashboard with Flux.

### More

- [X] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [X] Yes, I ran `pre-commit run -a` with this PR
- [X] Yes, I have added a new example under [examples](https://github.com/aws-observability/terraform-aws-observability-accelerator/tree/main/examples) to support my PR (when applicable)
- [X] Yes, I have updated the [Pages](https://github.com/aws-observability/terraform-aws-observability-accelerator/tree/main/docs) for this feature

**Note**: Not all the PRs required examples and docs.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

**Comment by bonclay7** [comment]
This looks good @elamaran11, how do we validate this?

**Comment by elamaran11** [comment]
> This looks good @elamaran11, how do we validate this?

@bonclay7 Just run it as is. It will work and you can validate a new dashboard getting created.

**Comment by bonclay7** [comment]
@elamaran11 To fully validate this, maybe we should drop [this in terraform?](https://github.com/aws-observability/terraform-aws-observability-accelerator/blob/main/modules/eks-monitoring/dashboards.tf#L31-L35) so we are sure the operator set it up correctly

**Comment by bonclay7** [comment]
So I went ahead and did it, I dropped the dashboard provided by Terraform but for some reason the operator can't renew my API key, maybe i'm doing something wrong? 

```
1.6857285045139527e+09	INFO	found matching Grafana instances for datasource	{"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"grafanadatasource-sample-amp","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "grafanadatasource-sample-amp", "reconcileID": "72cfd60c-a255-44a1-bfbd-88b0cbc4f90c", "count": 1}
1.6857285045445938e+09	INFO	found matching Grafana instances for dashboard	{"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"external-grafanadashboard-url","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "external-grafanadashboard-url", "reconcileID": "5a07e7b0-d200-4223-94bc-a2fe6e6009db", "count": 1}
1.6857285045556655e+09	ERROR	error reconciling datasource	{"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"grafanadatasource-sample-amp","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "grafanadatasource-sample-amp", "reconcileID": "72cfd60c-a255-44a1-bfbd-88b0cbc4f90c", "datasource": "grafanadatasource-sample-amp", "grafana": "external-grafana", "error": "status: 401, body: {\"message\":\"Expired API key\"}\n"}
github.com/grafana-operator/grafana-operator/controllers.(*GrafanaDatasourceReconciler).Reconcile
	github.com/grafana-operator/grafana-operator/controllers/datasource_controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:121
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:320
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234
1.6857285046102674e+09	ERROR	error reconciling dashboard	{"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"external-grafanadashboard-url","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "external-grafanadashboard-url", "reconcileID": "5a07e7b0-d200-4223-94bc-a2fe6e6009db", "dashboard": "external-grafanadashboard-url", "grafana": "external-grafana", "error": "status: 401, body: {\"message\":\"Expired API key\"}\n"}
github.com/grafana-operator/grafana-operator/controllers.(*GrafanaDashboardReconciler).Reconcile
	github.com/grafana-operator/grafana-operator/controllers/dashboard_controller.go:216
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:121
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:320
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234
```

**Comment by elamaran11** [comment]
> @elamaran11 To fully validate this, maybe we should drop [this in terraform?](https://github.com/aws-observability/terraform-aws-observability-accelerator/blob/main/modules/eks-monitoring/dashboards.tf#L31-L35) so we are sure the operator set it up correctly?
> 
> And [this too](https://github.com/aws-observability/terraform-aws-observability-accelerator/blob/main/modules/eks-monitoring/dashboards/nodeexporter-nodes.json)?

@bonclay7 Dropping is not required both boards can coexist for now.

**Comment by elamaran11** [comment]
> So I went ahead and did it, I dropped the dashboard provided by Terraform but for some reason the operator can't renew my API key, maybe i'm doing something wrong?
> 
> ```
> 1.6857285045139527e+09	INFO	found matching Grafana instances for datasource	{"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"grafanadatasource-sample-amp","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "grafanadatasource-sample-amp", "reconcileID": "72cfd60c-a255-44a1-bfbd-88b0cbc4f90c", "count": 1}
> 1.6857285045445938e+09	INFO	found matching Grafana instances for dashboard	{"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"external-grafanadashboard-url","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "external-grafanadashboard-url", "reconcileID": "5a07e7b0-d200-4223-94bc-a2fe6e6009db", "count": 1}
> 1.6857285045556655e+09	ERROR	error reconciling datasource	{"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"grafanadatasource-sample-amp","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "grafanadatasource-sample-amp", "reconcileID": "72cfd60c-a255-44a1-bfbd-88b0cbc4f90c", "datasource": "grafanadatasource-sample-amp", "grafana": "external-grafana", "error": "status: 401, body: {\"message\":\"Expired API key\"}\n"}
> github.com/grafana-operator/grafana-operator/controllers.(*GrafanaDatasourceReconciler).Reconcile
> 	github.com/grafana-operator/grafana-operator/controllers/datasource_controller.go:209
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:121
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:320
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234
> 1.6857285046102674e+09	ERROR	error reconciling dashboard	{"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"external-grafanadashboard-url","namespace":"grafana-operator"}, "namespace": "grafana-operator", "name": "external-grafanadashboard-url", "reconcileID": "5a07e7b0-d200-4223-94bc-a2fe6e6009db", "dashboard": "external-grafanadashboard-url", "grafana": "external-grafana", "error": "status: 401, body: {\"message\":\"Expired API key\"}\n"}
> github.com/grafana-operator/grafana-operator/controllers.(*GrafanaDashboardReconciler).Reconcile
> 	github.com/grafana-operator/grafana-operator/controllers/dashboard_controller.go:216
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:121
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:320
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273
> sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
> 	sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234
> ```

@bonclay7 The Grafana Key in the Secrets Manager is expired. The external secrets will take care of refreshing secrets manager key with k8s secret for grafana operator but operator has to take care of refereshing grafana keys with below :


```bash
export GO_AMG_API_KEY=$(aws grafana create-workspace-api-key \
  --key-name "grafana-operator-key" \
  --key-role "ADMIN" \
  --seconds-to-live 432000 \
  --workspace-id <YOUR_WORKSPACE_ID> \
  --query key \
  --output text)
  
export GO_API_KEY_SECRET_NAME=$(aws secretsmanager list-secrets \
    --query 'SecretList[1].Name' | tr -d '"')

aws secretsmanager update-secret \
    --secret-id $GO_API_KEY_SECRET_NAME \
    --secret-string "${GO_AMG_API_KEY}" \
    --region $GO_AWS_REGION
```

**Comment by bonclay7** [comment]
Did I miss these instructions in the docs, or we are waiting for the full move to make it clear?

**Comment by bonclay7** [comment]
> Dropping is not required both boards can coexist for now

Yes, but for customers using latest, they would be quite confused that we have two data sources and duplicates dashboards. I would suggest that we make it more seamless

**Comment by elamaran11** [comment]
> Did I miss these instructions in the docs, or we are waiting for the full move to make it clear?

This is same as how you upgrade the Grafana Keys today. We can add these instructions after full move

**Comment by elamaran11** [comment]
> > Dropping is not required both boards can coexist for now
> 
> Yes, but for customers using latest, they would be quite confused that we have two data sources and duplicates dashboards. I would suggest that we make it more seamless

Agreed. I commented out the NodeExporter in `dashboard.tf` but we can move it fully out in the full migration.

**Comment by bonclay7** [comment]
I can confirm after updating the secrets, the new DS pops up and Dashboard too. 

 I think it will be more sleek if we transition all the dashboards / folders at once so we keep the existing DS name `aws-observability-accelerator` and dashboard folders 
image

That way we have a transparent transition. But cool to see it works! happy to help on that if you need

**Comment by elamaran11** [comment]
> I can confirm after updating the secrets, the new DS pops up and Dashboard too.
> 
> I think it will be more sleek if we transition all the dashboards / folders at once so we keep the existing DS name `aws-observability-accelerator` and dashboard folders image
> 
> That way we have a transparent transition. But cool to see it works! happy to help on that if you need

Thankyou for confirming that Grafana Operator works to create Dashboards in AMG via GitOps. Only problem i have now is Folder creation is not working with external grafana. I have raised [this](https://github.com/grafana-operator/grafana-operator/issues/1079) issue with grafana-operator GitHub. Once folder works i can completely move it to new solution. I was thinking this as pilot. I will keep you posted on the issue. But let me know if you want to hold this PR or create all boards without folder or just have this board alone without folder. Im fine with either approaches

**Comment by elamaran11** [comment]
@bonclay7 Now i have all dashboards in the mix with Grafana Operator and commented out all dashboards in `dashboard.tf`. We are dependent on the GrafanaFolder feature from the ticket i previously mentioned and also dependent on this [PR](https://github.com/aws-observability/aws-observability-accelerator/pull/3).

**Comment by elamaran11** [comment]
@bonclay7 Grafana Folder works now with Grafana Operator. So i have migrated all the 6 dashboards for EKS Monitoring Module to Grafana Operator and Flux with GitOps. I think we can seamlessly move other pattern dashboards and cleanup as subsequent work in future PRs. Please take this forward and let me know if you have questions.

**Comment by elamaran11** [comment]
Added an operational procedure as well under `Troubleshooting` on how to replace the expired Grafana API Keys in AWS Secrets Manager.

**Comment by elamaran11** [comment]
image

image

File: modules/eks-monitoring/variables.tf [hunk 1]
```diff
@@ -345,6 +369,47 @@ variable "grafana_api_key" {
...
 
+variable "grafana_url" {
+  description = "Endpoint URL of Amazon Managed Grafana workspace"
+  type        = string
+}
+
+variable "grafana_cluster_dashboard_url" {
+  description = "Dashboard URL for Cluster Grafana Dashboard JSON"
+  type        = string
+  default     = "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/1d731aca31cdeb26e9fe9d017e609a5ba1621a30/artifacts/grafana-dashboards/cluster.json"
```
Comment (bonclay7) [inline modules/eks-monitoring/variables.tf:380]: This is awesome, it means customers can have their own dashboards in their own repos if they wish

**Review by bonclay7 — APPROVED** [review]
Awesome job @elamaran11. There is a few cleanup to do after this one, and docs updates

**Pull request merged**